mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala

```diff
@@ -316,24 +319,129 @@ class GaussianMixture @Since("2.0.0") (
   def setSeed(value: Long): this.type = set(seed, value)
 
+  // number of samples per cluster to use when initializing Gaussians
+  private val nSamples = 5
+
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): GaussianMixtureModel = {
     transformSchema(dataset.schema, logging = true)
-    val rdd: RDD[OldVector] = dataset.select(col($(featuresCol))).rdd.map {
-      case Row(point: Vector) => OldVectors.fromML(point)
+
+    val sc = dataset.sparkSession.sparkContext
+    val _k = $(k)
+
+    val instances: RDD[Vector] = dataset.select(col($(featuresCol))).rdd.map {
+      case Row(features: Vector) => features
+    }.cache()
+
+    // Extract the number of features.
+    val numFeatures = instances.first().size
+
+    val shouldDistributeGaussians = GaussianMixture.shouldDistributeGaussians(_k, numFeatures)
+
+    // Determine initial weights and corresponding Gaussians.
+    // We start with uniform weights, a random mean from the data, and
+    // diagonal covariance matrices using component variances
+    // derived from the samples.
+    // TODO: Support users supplied initial GMM.
+    val samples = instances.takeSample(withReplacement = true, _k * nSamples, $(seed))
+    val weights: Array[Double] = Array.fill(_k)(1.0 / _k)
+    /**
+     * Since the covariance matrix of multivariate gaussian distribution is symmetric,
+     * only the upper triangular part of the matrix will be stored as a dense vector
+     * in order to reduce the shuffled data size.
+     */
+    val gaussians: Array[(DenseVector, DenseVector)] = Array.tabulate(_k) { i =>
+      val slice = samples.view(i * nSamples, (i + 1) * nSamples)
+      val mean = {
+        val v = new DenseVector(Array.fill[Double](numFeatures)(0.0))
+        var i = 0
+        while (i < nSamples) {
+          BLAS.axpy(1.0, slice(i), v)
+          i += 1
+        }
+        BLAS.scal(1.0 / nSamples, v)
+        v
+      }
+      /**
+       * Construct matrix where diagonal entries are element-wise
+       * variance of input vectors (computes biased variance).
+       */
+      val cov = {
+        val ss = new DenseVector(Array.fill[Double](numFeatures)(0)).asBreeze
+        slice.foreach(xi => ss += (xi.asBreeze - mean.asBreeze) :^ 2.0)
+        val diagVec = Vectors.fromBreeze(ss)
+        BLAS.scal(1.0 / nSamples, diagVec)
+        val covVec = new DenseVector(Array.fill[Double](
+          numFeatures * (numFeatures + 1) / 2)(0.0))
+        diagVec.toArray.zipWithIndex.foreach { case (v: Double, i: Int) =>
+          covVec.values(i + i * (i + 1) / 2) = v
+        }
+        covVec
+      }
+      (mean, cov)
     }
 
-    val algo = new MLlibGM()
-      .setK($(k))
-      .setMaxIterations($(maxIter))
-      .setSeed($(seed))
-      .setConvergenceTol($(tol))
-    val parentModel = algo.run(rdd)
-    val gaussians = parentModel.gaussians.map { case g =>
-      new MultivariateGaussian(g.mu.asML, g.sigma.asML)
+    var llh = Double.MinValue // current log-likelihood
```

Comment:
minor: why not `logLikelihood` and `logLikelihoodPrev` ? It's nice to have descriptive variable names, then we can remove the comments.